.github/workflows/ubuntu.yml

```diff
@@ -0,0 +1,128 @@
+          - name: "Build Focal"
+            release: focal
+    
+          - name: "Build Groovy"
+            release: groovy
+    
+    steps:
+
+      - name: Check Out Repo 
+        uses: actions/checkout@v2
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_TOKEN }}
+          
+      - name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: docker/actions/Dockerfile.actions.${{ matrix.release }}
+          push: true
+          tags: fluidity/actions:${{ matrix.release }}-${{ github.sha }}
+
+  testing:
+
+    name: ${{ matrix.name }}
+    runs-on: ubuntu-latest
+    needs: build
+    if: always()
+
+    strategy:
+      fail-fast: false
+      matrix:
+        name: [ "Unit Bionic", "Short Bionic", "Medium Bionic", "Unit Focal", "Short Focal", "Medium Focal", "Unit Groovy", "Short Groovy", "Medium Groovy" ]
+        include:
+        
+          - name: "Unit Bionic"
+            release: bionic
+            command: "make unittest"
+            output: "test_results_unittests.xml"
+          
+          - name: "Short Bionic"
+            release: bionic
+            command: "make THREADS=2 test"
+            output: "test_results.xml"
+
+          - name: "Medium Bionic"
+            release: bionic
+            command: "make THREADS=2 mediumtest"
+            output: "test_results_medium.xml"
+
+          - name: "Unit Focal"
+            release: focal
+            command: "make unittest"
+            output: "test_results_unittests.xml"
+          
+          - name: "Short Focal"
+            release: focal
+            command: "make THREADS=2 test"
+            output: "test_results.xml"
+
+          - name: "Medium Focal"
+            release: focal
+            command: "make THREADS=2 mediumtest"
+            output: "test_results_medium.xml"
+
+          - name: "Unit Groovy"
+            release: groovy
+            command: "make unittest"
+            output: "test_results_unittests.xml"
+          
+          - name: "Short Groovy"
+            release: groovy
+            command: "make THREADS=2 test"
+            output: "test_results.xml"
+
+          - name: "Medium Groovy"
+            release: groovy
+            command: "make THREADS=2 mediumtest"
+            output: "test_results_medium.xml"
+
+    steps:
+
+      - name: ${{ matrix.name }} Testing
+        run: |
+         chmod 777 .
+         docker pull fluidity/actions:${{ matrix.release }}-${{ github.sha }}
+         docker run -v $PWD:/host fluidity/actions:${{ matrix.release }}-${{ github.sha }} /bin/bash -c "${{ matrix.command }} && cp -v tests/${{ matrix.output }} /host/${{ matrix.release }}-${{ matrix.output}}"
+
+      - uses: actions/upload-artifact@v2
+        with:
+          path: ${{ matrix.release }}-${{ matrix.output }}
```

Comment:
You could potentially add `name: tests_xml_outputs` to make it more obvious what this "artifact" archive contains.

Reply:
Good catch! I'd missed that that was lacking a name - fixed in ef3a9ba .